### DISCUSSION TITLE
Project start-up

### DISCUSSION THREAD

**Discussion by JamesSample** [NIVANorge/lake_liming_app#4] [General]
@nmarquesin @ivanaH8

Thanks for your interest in developing the lake liming app! The contract is now signed and I have a project code, so we can get started as soon as you're ready :-) I've done some initial "bemanning" in Instipro, but I'll adjust this as necessary. I don't know how much time and interest you both have, but it's not a big project so we need to keep things quite focused.

I've spent a few hours this morning tidying up the repository on GitHub. Please take a look at the links below and let me know if you have any questions or comments.

- The [Wiki](https://github.com/NIVANorge/lake_liming_app/wiki) page includes an overview of the project, an approximate timeline and the main development tasks (details [here](https://github.com/NIVANorge/lake_liming_app/wiki/Development-tasks)). You can also find the NIVA project code here.

- I thought we could use [Quarto](https://quarto.org/docs/guide/) for the documentation. **Please let me know if this is not OK**. This basically involves having `.qmd` files within the `docs` subfolder. Whenever you edit these files and push changes to `main`, the GitHub action [here](https://github.com/NIVANorge/lake_liming_app/blob/main/.github/workflows/quarto-publish.yml) will re-render the pages and update the website [here](https://nivanorge.github.io/lake_liming_app/).

- I've created a simple Streamlit app [here](https://nivanorge-lake-liming-app-app-ywryvk.streamlit.app/). If you prefer, **I'm happy for the final app to be implemented in JavaScript, or anything else**. I created the Streamlit version because I want to have something "hands-on" to discuss with Helge and Karl-Jan, and I find it almost as fast to develop examples with Streamlit as to create mock-ups using e.g. PowerPoint or Illustrator. 

I think the first steps are probably as follows:

1. **James** - Contact Karl-Jan and Helge with the proposed Excel template for the column tests and a link to the Streamlit app. I'm pretty sure my template needs generalising to allow for some additional options. 

2. **Nathalie** - Consider the pros and cons of using Streamlit vs. e.g. JavaScript. If you think a JavaScript solution will be better/more maintainable, and achievable within the project budget, then I'm happy to go forward with that. 

Thanks!

**Comment by nmarquesin** [comment]
@JamesSample, after going through your code, the Streamlit docs, and the hours available for this project, I don't think I can create a js app that will be better/more maintainable. 
A js app using graphing libraries would have many more lines of code in comparison. I rewrote parts of your code using [plotly.js](https://plotly.com/javascript/), and [chart.js](https://www.chartjs.org/docs/latest/), and even tried it as a flask app with plotly for python. Your code using Streamlit is always much shorter.   
It is justifiable to go this js/python-flask route if there's a project requirement that cannot be met because of a Streamlit limitation. Are you anticipating any? Is there something you want to do in this app that Streamlit doesn't cater for? 
Another possible reason to go with a js app would be if we already had similar code ready to be reused.
Other than these reasons, I can't really propose we go with a different technology for this project.

**Reply by IvanaH8** [reply]
I agree with @nmarquesin. Although I do not have much experience with either technologies, after little research it seems like Streamlit will indeed be easier and quicker to implement for this purpose, taking into account background calculations, visualization capabilities, number of users and access control and, as Nathalie mentioned, the complexity of the code. 
 
I'm looking forward to dive into it either way!

**Reply by JamesSample** [reply]
Thanks for the feedback, @nmarquesin and @IvanaH8!

My experience of developing Streamlit apps has mostly been good, although they have a tendency to start off nice and clean, and then become increasingly "hacky" as the project progresses.

I'm reasonably sure that Streamlit will cater for everything we need in this project. The most tricky parts I am anticipating are (i) the lake modelling, and (ii) generating summary PDFs for users to download. Both of these are essentially independent of the choice of UI library, and I have previous experience of doing both in Python. I think the UI just needs to display some plots and tables, which Streamlit does OK.

My only knowledge of developing UIs is using "tools-for-people-that-can't-code-UIs" (i.e. Streamlit, Shiny, Dash etc.). I therefore don't have much appreciation of what would be involved in a custom solution using e.g. JavaScript. From my perspective, **the most important thing is that you're both happy with whatever tools we decide to use**, since I'm hoping you'll share most of the development work between you. 

So, continuing with Streamlit sounds good to me (but don't feel you have to just because it's all I can do) :-)

I had some feedback yesterday from Karl-Jan regarding the column test calculations. Once I've clarified a few more things I'll create some issues and maybe start a new discussion thread for what happens next.

Thanks!
